### ISSUE TITLE
Licence Store

### ISSUE THREAD

**Issue by jbonnet** [sonata-nfv/son-gkeeper#466]
We need to build a very simple 'Licence Store', where

* Customers can buy service instantiation licences;
* Developers can:
  * buy package download licences (i.e., for service/function re-use);
  * list current licences of his/her services

@srodriguezOPT: do you think we could include it in the BSS? Or @trakadasp it fits better within the GUI? Or elsewhere?

This would improve demonstration capability of the Private licencing

**Comment by srodriguezOPT** [comment]
Hi,

BSS includes a "get licence" invocation to allow the instantiation of non public services, but i think that we should use a centralized component (e.g the gatekeeper) to manage the licences because it will be invoked since GUI and BSS.

On the other hand, think that BSS is an external component that could be changed by a third party bss that invokes the gatekeeper and without our sonata licence management.

**Comment by jbonnet** [comment]
@srodriguezOPT 
You're right, we'll be using the GK to 'manage' (whatever that means) licences, but we need a GUI to list, buy, etc., no?

**Comment by srodriguezOPT** [comment]
Yes, 
it's needed to define this license management and what we will achieve because it will be as complicated as we want. E.g, in the BSS:

- a unique licence that allows to instantiate (1..N instances), update, ... all actions allowed

vs

- licence to instantiate 1 instance
- licence to instantiate multiple instances
- licence to update an instance
- ...

and GUI is even more complex (more use cases)

Currently, BSS only consider service instantiation licences. If the user can instantiate (has licence) the private service, the instantiation button appears; otherwise, the "get licence" appears:

![image](https://cloud.githubusercontent.com/assets/15909508/24350416/6c13109c-12e2-11e7-9150-ba6e08a53e3a.png)

**Comment by srodriguezOPT** [comment]
Hi, @jbonnet 

From BSS side, I was thinking in maintain the get license button in the available services menu (getting service license in this tab implies to obtain service and functions instantiation licenses).

On the other hand we can create a new "License Store" menu differentiating between service and function licenses with two tabs (I think we need to manage function licenses because developers could require function licenses to compose their packages and could be not interested in obtain the other functions licenses that compose the original service).

The Service/Function licenses tab could be like:

| Name | Vendor | Version | Purpose | type |
|:--------|:---------|:---------|:----------|:------------------------|
| name1 | vendor1 | version | instantiation | public |
| name2 | vendor2 | version | development | private |

@dang03 told me that there are three types of user roles: 
- developers
- customers
- both

We can show the information based on the logged in user role: customer only sees the instantiation purpose licenses, developer only sees the development purpose licenses and users with both types can see all information.

We will need:
- to retrieve functions information (GK -> BSS)
- to retrieve the user role (GK -> BSS)
- to define the get licenses request: one by one? multiple requests (e.i with checkboxes including all requests in one json)? (BSS -> GK)
- to decompose get service license in get service and functions licenses (GK)
[...]

What do you think?

**Comment by jbonnet** [comment]
@srodriguezOPT 
Let's assume the simplest case for instantiation: one licence -> multiple instances allowed

**Comment by srodriguezOPT** [comment]
@jbonnet 

OK. So, with this simplification, function licenses tab is not needed. We will work at service level: service license implies service and all its functions licenses.

What happen if a developer needs to include a function and he doesn't have the service's license that contains that function? Does he need to get the other service license first? Does it out of the scope?

**Comment by srodriguezOPT** [comment]
Hi @jbonnet,

The BSS' license store development is done. Currently, the interaction with the gatekeeper is mocked in this way:

- **GET** user licenses:
$http.get(ENV.apiEndpoint+"**/licenses/user/**"+$localStorage.currentUser.user_id+"?limit="+10+"&offset="+offset)
expecting something like:
   [
	{
        "service_name": "sonata-demo-1",
        "service_id": "8c58b169-7c38-4bcd-9421-a91bd786f1fa",
        "license_type": "Public",  or "Private"
        "license_use": "Instantiation" or "Package Creation"
         },
         {...}
   ]
   (pagination links included)

- **POST** license (license requested by user):
$http.post(ENV.apiEndpoint+"**/licenses**",data)
where data={"service_uuid":id, "user_id":$localStorage.currentUser.user_id, "user_role":$localStorage.currentUser.user_role};
